src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -0,0 +1,158 @@
+package fr.acinq.lightning.wire
+
+import fr.acinq.bitcoin.*
+import fr.acinq.bitcoin.io.ByteArrayOutput
+import fr.acinq.bitcoin.io.Input
+import fr.acinq.bitcoin.io.Output
+import fr.acinq.lightning.MilliSatoshi
+import fr.acinq.lightning.blockchain.fee.FeeratePerKw
+import fr.acinq.lightning.channel.*
+import fr.acinq.lightning.transactions.Transactions
+import fr.acinq.lightning.utils.Either
+import fr.acinq.lightning.utils.msat
+import fr.acinq.lightning.utils.sat
+
+/**
+ * Liquidity ads create a decentralized market for channel liquidity.
+ * Nodes advertise fee rates for their available liquidity using the gossip protocol.
+ * Other nodes can then pay the advertised rate to get inbound liquidity allocated towards them.
+ */
+object LiquidityAds {
+
+    /**
+     * @param miningFee fee paid to miners for the underlying on-chain transaction.
+     * @param serviceFee fee paid to the liquidity provider for the inbound liquidity.
+     */
+    data class LeaseFees(val miningFee: Satoshi, val serviceFee: Satoshi) {
+        val total: Satoshi = miningFee + serviceFee
+    }
+
+    /**
+     * Liquidity is leased using the following rates:
+     *
+     *  - the buyer pays [leaseFeeBase] regardless of the amount contributed by the seller
+     *  - the buyer pays [leaseFeeProportional] (expressed in basis points) of the amount contributed by the seller
+     *  - the seller will have to add inputs/outputs to the transaction and pay on-chain fees for them, but the buyer
+     *    refunds on-chain fees for [fundingWeight] vbytes
+     *
+     * The seller promises that their relay fees towards the buyer will never exceed [maxRelayFeeBase] and [maxRelayFeeProportional].
+     * This cannot be enforced, but if the buyer notices the seller cheating, they should blacklist them and can prove
+     * that they misbehaved using the seller's signature of the [LeaseWitness].
+     */
+    data class LeaseRate(val leaseDuration: Int, val fundingWeight: Int, val leaseFeeProportional: Int, val leaseFeeBase: Satoshi, val maxRelayFeeProportional: Int, val maxRelayFeeBase: MilliSatoshi) {
```

Comment:
> Nit: traditionally, base is before prop:

Lisa chose to include the proportional fees first in the spec messages, that's why I kept it that way (because in eclair we'll have to follow the encoding order, and it's nice to be consistent with eclair).

> Also, prop units are misleading: leaseFeeProportional is in per-10,000, maxRelayFeeProportional is in per-1,000,000.

Here, they are both basis points for consistency (which I agree is confusing, but there has to be confusion somewhere while we're moving to basis points instead of ppm). I can rename them to `ProportionalBasisPoint` to clarify though!

Reply:
Noted thanks!

Okay, so the convention is that "proportional" is bips by default, I think that's fine.